### PR TITLE
Permissions support

### PR DESCRIPTION
**Pull request by sanjana-sfdc**

### CODE REVIEW

File: argusclient/client.py [hunk 1]
```diff
@@ -438,13 +439,50 @@ def _init_all(self, coll=None):
...
         """
         Gets permissions that are associated with the given entity id's.
 
         :return: a dict of entity id's mapped to a list of :class:`argusclient.model.Permission` objects
         """
-        return convert(self.argus._request("post", "permission/entityIds", dataObj=entityIds))
+        entityIds = []
+        for entity_id in entity_ids:
+            if entity_id not in self._coll:
+                entityIds.append(entity_id)
```
Comment (j-ma-sf) [inline argusclient/client.py:451]: Is it filtering out entity id's for which the permissions have not yet been queried? To save processing time?

File: tests/test_data.py [hunk 2]
```diff
@@ -38,6 +38,12 @@
 userPermissionIdentifier = "user"
 permissionNames = ["VIEW", "EDIT", "DELETE"]
 permissionGroupId = '24231-52321-43523-64353-23111'
+username = "testuser"
+permission_ids = [1,2,3]
+user_type = "user"
+group_type = "group"
```
Comment (j-ma-sf) [inline tests/test_data.py:44]: I think either this or userPermissionIdentifier and groupPermissionIdentifier can be used

File: argusclient/client.py [hunk 3]
```diff
@@ -438,13 +439,50 @@ def _init_all(self, coll=None):
...
+        if updated_permission.entityId != entity_id:
+            raise ArgusException("This is unexpected... permission: %s not associated with entity after successfully"
+                                 " adding it" % permission)
+        return updated_permission
+
+    def delete(self, entity_id, permission):
+        if not isinstance(permission, Permission):
+            raise TypeError("Need a Permission object, got: %s" % type(permission))
+        if permission.type == "user" and permission.permissionIds == []:
+            raise ValueError("A user permission needs to have the permission that needs to be revoked")
```
Comment (j-ma-sf) [inline argusclient/client.py:481]:
```suggestion
            raise ValueError("Deleting a user permission requires the permission that needs to be revoked")
```

File: argusclient/client.py (hunk 1 above)
Comment (sanjana-sfdc) [reply argusclient/client.py:451]: yeah and not make the request again if we already have the info from a previous call for that entity ID made

File: argusclient/client.py (hunk 3 above)
Comment (sanjana-sfdc) [reply argusclient/client.py]: Done

File: argusclient/model.py [hunk 4]
```diff
@@ -262,11 +262,12 @@ class Permission(BaseEncodable):
...
 
-    id_fields = ("permissionNames",)
-
-    def __init__(self, type, permissionNames, **kwargs):
-        super(Permission, self).__init__(type=type, permissionNames=permissionNames, **kwargs)
+    id_fields = ("type",)
+    VALID_TYPES = frozenset(("user", "group"))
 
+    def __init__(self, type, **kwargs):
+        assert type in Permission.VALID_TYPES, "permission type is not valid: %s" % type
```
[the remark this answers is not in the document]
Comment (sanjana-sfdc) [reply argusclient/model.py]: Done

File: tests/test_data.py (hunk 2 above)
Comment (sanjana-sfdc) [reply tests/test_data.py]: Done

File: argusclient/client.py [hunk 5]
```diff
@@ -438,13 +439,50 @@ def _init_all(self, coll=None):
...
 
-    def get_permissions_for_entities(self, entityIds):
+    def get_permissions_for_entities(self, entity_ids):
         """
         Gets permissions that are associated with the given entity id's.
 
         :return: a dict of entity id's mapped to a list of :class:`argusclient.model.Permission` objects
         """
-        return convert(self.argus._request("post", "permission/entityIds", dataObj=entityIds))
+        entityIds = []
```
Comment (j-ma-sf) [inline argusclient/client.py:448]:
```suggestion
        # Filter out entity id's for which the permissions have already been queried
        entityIds = []
```

File: argusclient/client.py (hunk 1 above)
Comment (j-ma-sf) [reply argusclient/client.py:451]: Ah ok

**Review by j-ma-sf — APPROVED** [review]

File: argusclient/client.py (hunk 5 above)
Comment (sanjana-sfdc) [reply argusclient/client.py:448]: done

**Pull request merged**